### PR TITLE
Added `==` and `!=` operator overloads to `IpAddress` 

### PR DESCRIPTION
**Pull request by CoryCharlton**
## Description
Added `==` and `!=` operator overloads

## Motivation and Context
There is existing code that does not use the Equals method so this will fix that.

## How Has This Been Tested?
On device

## Types of changes

- [X] Improvement (non-breaking change that improves a feature, code or algorithm)
- [ ] Bug fix (non-breaking change which fixes an issue with code or algorithm)
- [X] New feature (non-breaking change which adds functionality to code)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Config and build (change in the configuration and build system, has no impact on code or features)
- [ ] Dependencies (update dependencies and changes associated, has no impact on code or features)
- [ ] Unit Tests (add new Unit Test(s) or improved existing one(s), has no impact on code or features)
- [ ] Documentation (changes or updates in the documentation, has no impact on code or features)

## Checklist:

- [X] My code follows the code style of this project (only if there are changes in source code).
- [x] My changes require an update to the documentation (there are changes that require the docs website to be updated).
- [ ] I have updated the documentation accordingly (the changes require an update on the docs in this repo).
- [X] I have read the [CONTRIBUTING](https://github.com/nanoframework/.github/blob/main/CONTRIBUTING.md) document.
- [X] I have tested everything locally and all new and existing tests passed (only if there are changes in source code).
- [x] I have added new tests to cover my changes.

### CODE REVIEW

**Review by MrCSharp22 — APPROVED** [review]
Great improvement. Looks good to me. Thank you!

**Comment by josesimoes** [comment]
Interestingly, we are ahead of .NET!! 🤩  
Take a look at dotnet/runtime#70978

File: Tests/IPAddressTests/IPAddressTests.cs [hunk 1]
```diff
@@ -301,5 +301,24 @@ public void NetTest6_SocketAddressBasic()
...
         }
+
+        [TestMethod]
+        public void Equality_Tests()
+        {
+            var privateAddress = new IPAddress(new byte[] { 192, 168, 0, 1 });
+            var publicAddress = new IPAddress(new byte[] { 1, 1, 1, 1 });
+
+            // Equal
+            Assert.IsTrue(privateAddress.Equals(new IPAddress(new byte[]{192, 168, 0 , 1})), "192.168.0.1 equals 192.168.0.1");
```
Comment (josesimoes) [inline Tests/IPAddressTests/IPAddressTests.cs:312]: We should get away of "magic" numbers as they are prone to errors.
Better move these byte arrays for the addresses to readonly fields and/or having IP addresses as strings and then use Parse.

File: Tests/IPAddressTests/IPAddressTests.cs [hunk 2]
```diff
@@ -301,5 +301,24 @@ public void NetTest6_SocketAddressBasic()
...
+            Assert.IsTrue(publicAddress.Equals(new IPAddress(new byte[] { 1, 1, 1, 1 })), "1.1.1.1 equals 1.1.1.1");
```
Comment (josesimoes) [inline Tests/IPAddressTests/IPAddressTests.cs:313]: The comment you've made are rather similar to the ones that the test framework produces by default. Just mentioning it so you don't have all this trouble in future work.

**Comment by josesimoes** [comment]
On Sonarcloud bug: I get their point and it makes sense. Looking at the code, there is no major calculation going on with generating the hashcode, so we're probably good with dropping the _hash field and simple compute the hash when needed. 
Thoughts @CoryCharlton ?

File: Tests/IPAddressTests/IPAddressTests.cs (hunk 2 above)
Comment (CoryCharlton) [reply Tests/IPAddressTests/IPAddressTests.cs:313]: I think I added these because `IsTrue` doesn't have a message by default. Although in this case I could have used `AreEqual` as I actually want to use the `Equals` method (I think it was copy/paste from the `==` cases). 

Will update.

File: Tests/IPAddressTests/IPAddressTests.cs (hunk 1 above)
Comment (CoryCharlton) [reply Tests/IPAddressTests/IPAddressTests.cs]: Switch to `IPAddress.Parse`. Last I checked class level fields don't work because the test class isn't constructed (at least this was the case with the `SetupAttribute`.

**Comment by CoryCharlton** [comment]
> On Sonarcloud bug: I get their point and it makes sense. Looking at the code, there is no major calculation going on with generating the hashcode, so we're probably good with dropping the _hash field and simple compute the hash when needed. Thoughts @CoryCharlton ?

I can't see the error now because workflow is re-running but I assume it was something along the lines of "non-readonly field used in GetHashCode" as Resharper gave me the same warning.

I was copying the pattern I saw in the .NET full implementation here. They are doing slightly more work but I think this is also a case of "maximum optimization" as no work is better than some work in a method like `GetHashCode` where you expect it to be as fast as possible. 

That being said I don't have any objections to removing the field and calculating every time so I'll leave it up to you.

**Comment by josesimoes** [comment]
That's exactly the same complain, yes. 
Let's address the warning and recalculate each time. I'm not seeing a big need to cache the hash value instead of recalculating when needed.

File: Tests/IPAddressTests/IPAddressTests.cs (hunk 1 above)
Comment (josesimoes) [reply Tests/IPAddressTests/IPAddressTests.cs]: You're correct on that!
But static or const do work, right? (can't recall.... 😅 )

**Review by Ellerbach — COMMENTED** [review]
Stupid question: who will this play with the larger PR opened by @AdrianSoundy which also touches this?
Shall we wait a bit for the larger one to go and then have this one? 
@josesimoes , any thoughts as well? (I know you want to add as well on this one)

**Comment by josesimoes** [comment]
> Stupid question: who will this play with the larger PR opened by @AdrianSoundy which also touches this? 

Not stupid at all! It makes sense to have #309 merged first as the merge into this one will be smoother, if not automatic.

**Comment by AdrianSoundy** [comment]
@CoryCharlton 
Should be able to complete this one now as other PR has been merged.

**Comment by CoryCharlton** [comment]
@AdrianSoundy @josesimoes I merged in the latest changes and think we should be good to go with this. There were several other changes Resharper was recommending to me (switching to var, auto properties, naming [ie: scopeid to scopeId], etc.) but I left them out to not make this change more complicated than it needs to be.

**Comment by josesimoes** [comment]
@CoryCharlton you did well. PRs need to be atomic. Not good adding other changes, even tiny/random fixes, code style improvements, etc.

File: nanoFramework.System.Net/IPAddress.cs [hunk 3]
```diff
@@ -24,12 +24,38 @@ public class IPAddress
         /// <summary>
         /// Provides an IP address that indicates that the server must listen for client activity on all network interfaces. This field is read-only.
         /// </summary>
-        public static readonly IPAddress Any = new(0x0000000000000000);
+        /// <remarks>
+        /// The <see cref="Socket.Bind"/> method uses the <see cref="Any"/> field to indicate that a <see cref="Socket"/> instance must listen for client activity on all network interfaces.
+        /// 
+        /// The <see cref="Any"/> field is equivalent to 0.0.0.0 in dotted-quad notation.
+        /// </remarks>
+        public static readonly IPAddress Any = new(new byte[] { 0, 0, 0, 0 });
```
Comment (josesimoes) [inline nanoFramework.System.Net/IPAddress.cs:32]: Why this change? The byte[] constructor is soooooo much more inefficient. 🤔

File: nanoFramework.System.Net/IPAddress.cs [hunk 4]
```diff
@@ -24,12 +24,38 @@ public class IPAddress
...
 
         /// <summary>
         /// Provides the IP loopback address. This field is read-only.
         /// </summary>
-        public static readonly IPAddress Loopback = new(0x000000000100007F);
+        /// <remarks>
+        /// The <see cref="Loopback"/> field is equivalent to 127.0.0.1 in dotted-quad notation.
+        /// </remarks>
+        public static readonly IPAddress Loopback = new(new byte[] { 127, 0, 0, 1 });
```
Comment (josesimoes) [inline nanoFramework.System.Net/IPAddress.cs:40]: Same question here

File: nanoFramework.System.Net/IPAddress.cs [hunk 5]
```diff
@@ -24,12 +24,38 @@ public class IPAddress
...
+
+        /// <summary>
+        /// Provides the IP broadcast address. This field is read-only.
+        /// </summary>
+        /// <remarks>
+        /// The <see cref="Broadcast"/> field is equivalent to 255.255.255.255 in dotted-quad notation.
+        /// </remarks>
+        public static readonly IPAddress Broadcast = new(new byte[] { 255, 255, 255, 255 });
```
Comment (josesimoes) [inline nanoFramework.System.Net/IPAddress.cs:48]: Same applies here

File: nanoFramework.System.Net/NetworkInformation/NetworkInterface.cs [hunk 6]
```diff
@@ -470,10 +470,10 @@ public NetworkInterfaceType NetworkInterfaceType
         #region native methods
 
         [MethodImpl(MethodImplOptions.InternalCall)]
-        private extern static int GetNetworkInterfaceCount();
+        private static extern int GetNetworkInterfaceCount();
```
Comment (josesimoes) [inline nanoFramework.System.Net/NetworkInformation/NetworkInterface.cs:473]: Please leave these change for another PR as they are unrelated.

File: nanoFramework.System.Net/IPAddress.cs (hunk 3 above)
Comment (CoryCharlton) [reply nanoFramework.System.Net/IPAddress.cs:32]: I thought we had a discussion earlier regarding avoiding magic numbers and came to the conclusion that the byte[] constructor or even the "Parse" method were clearer. 

Since these are all static readonly I'm not sure the efficiency is much concern.

Comment (josesimoes) [reply nanoFramework.System.Net/IPAddress.cs:32]: Forgive me but I can't recall that conversation... 😅
Anyway, both approaches are declaring the IP address to be used in the code, so that these are made available and no magic numbers are used in the code. 
At the declaration level, these numbers will always have to exist right? Either as byte array elements or integers...

BTW, most code, as usually happens was taken/inspired from the source code of the full .NET:
https://referencesource.microsoft.com/#System/net/System/Net/IPAddress.cs,18

Comment (CoryCharlton) [reply nanoFramework.System.Net/IPAddress.cs:32]: .NET Core went the other way any used the byte array constructors: https://github.com/dotnet/runtime/blob/342951dec5dc326a4b0ba0d0949b7a4155154e90/src/libraries/System.Net.Primitives/src/System/Net/IPAddress.cs#L20

I'm fine either way just let me know what we prefer.

Comment (josesimoes) [reply nanoFramework.System.Net/IPAddress.cs:32]: Interesting... well, nano code base is comming from 4.x so no surprise this was using that approach.
No need to put more work on this. OK with byte arrays. I was just checking why the change.

**Review by josesimoes — APPROVED** [review]
LGTM. Thanks for this improvement.

**Comment by Ellerbach** [comment]
@CoryCharlton I think there is an issue.
When doing:

```csharp
IPAddress ip = default
if( ip == default)
{
  // Should be true but it's not!
}
```

```csharp
IPAddress ip = null
if( ip == null)
{
  // Should be true but it's not!
}
```

**Comment by CoryCharlton** [comment]
@Ellerbach I'll take a look in a bit, should be a simple fix. In the meantime using the `is` operator will work as it doesn't use operators defined in the class.

**Comment by CoryCharlton** [comment]
@josesimoes / @Ellerbach fixed in https://github.com/nanoframework/System.Net/pull/311
